metrics/metrics.go

```diff
@@ -0,0 +1,447 @@
+package metrics
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	gm "github.com/rcrowley/go-metrics"
+	"github.com/square/etre"
+)
+
+const (
+	// Query counter is the grand total number of queries to the server.
+	// It can be used to calculate overall QPS.
+	Query byte = iota
+
+	// SetOp counter is the total number of queries that used a set op.
+	SetOp
+
+	// Labels gauge is the number of labels used by a query. The _min, _max,
+	// _avg, and _med (median) stats are reported for Labels.
+	Labels
+
+	// LatencyMs gauge is response time (in milliseconds) of a query. The _max,
+	// _p99 (99th percentile/top 1%), and _p999 (99.9th percentile/top 0.1%)
+	// stats are reported for LatencyMs.
+	LatencyMs
+
+	// MissSLA counter is the total number of queries with LatencyMs
+	// greater than the configured query latency SLA.
+	MissSLA
+
+	// Read counter is the grand total number of read queries to the server.
+	Read
+
+	// ReadQuery counter is the total number of read queries that selected
+	// entities with a query string (e.g. "host=foo,env=production").
+	// GET /entities and POST /query
+	ReadQuery
+
+	// ReadId counter is the total number of read queries that selected one
+	// entity by its unique _id.
+	// GET /entity/:id
+	ReadId
+
+	// ReadLabels counter is the total number of read label queries.
+	// GET /entity/:id/:label
+	ReadLabels
+
+	// Write counter is the grand total number of write queries to the server.
+	// PUT, POST, DELETE * (except ^)
+	Write
+
+	// POST /entity
+	Insert
+
+	// POST /entities
+	InsertBulk
+
+	// PUT /entity/:id
+	Update
+
+	// PUT /entities?query
+	UpdateBulk
+
+	// DELETE /entity/:id
+	Delete
+
+	// DELETE /entities?query
+	DeleteBulk
+
+	// DELETE /entity/:id/labels/:label
+	DeleteLabel
+
+	LabelRead
+	LabelUpdate
+	LabelDelete
+
+	DbError
+	APIError
+	ClientError
+
+	CDCClients
+)
+
+type Metrics interface {
+	// EntityType binds the Metrics instance to an entity type before calling
+	// other methods. This method must be called first.
+	EntityType(string)
+
+	// Inc increments the metric name (mn) by n. The metric must be a counter.
+	Inc(mn byte, n int64)
+
+	// IncLabel increments the metric name (mn) counter for the label by 1.
+	// The metric must be LabelRead, LabelUpdate, or LabelDelete.
+	IncLabel(mn byte, label string)
+
+	// IncError increments the metric name (mn) by 1. The metirc must be one
+	// of the *Error metrics.
+	IncError(mn byte)
+
+	// Val records one measurement (n) for the metric name (mn). The metric
+	// must be a gauge.
+	Val(mn byte, n int64)
+
+	// Trace increments trace value counters by 1.
+	Trace(map[string]string)
+
+	// Report returns a snapshot of all metrics, calculating stats like average
+	// and percentiles.
+	Report() etre.MetricsReport
+}
+
+var _ = &metrics{} // ensure metrics implements Metrics
```

Comment:
I've seen this several times now, but I don't understand - how does this ensure metrics implements Metrics? Wouldn't it need to be something like `var _ (Metrics) = &metrics{}`?

Reply:
Yes, you're correct. Fixed.